core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+
+    """
+
+    def __init__(self, config_manager, communication_manager):
+        """
+        TODO: Refactor dependencies
+        TODO: deal with config
+        """
+        # TODO: `communication_manager` is only used in a subset of methods,
+        # consider separating
+        self.communication_manager = communication_manager
+        # TODO: We will do this once the Communication Manager is done.
+        # self.communication_manager.configure_listener(self)
+
+        config = config_manager.get_config() if config_manager else {}
+        self.state = []
+        self.host = config.get("BLOCKCHAIN", "host")
+        self.ipfs_port = config.get("BLOCKCHAIN", "ipfs_port")
+        self.port = config.get("BLOCKCHAIN", "http_port")
+        self.client = None
+        try:
+            self.client = ipfsapi.connect(self.host, self.ipfs_port)
+        except Exception as e:
+            logging.info("IPFS daemon not started, got: {0}".format(e))
+
+        self.CALLBACKS = {
+            MessageEventTypes.NEW_WEIGHTS.name: self.broadcast_new_weights, 
+            MessageEventTypes.NOTHING.name: self._do_nothing,
+        }
+
+    ##########################################################################
+    ###                            API SECTION                             ###
+    ##########################################################################
+
+    def _do_nothing(self, payload: dict):
+        """
+        Do nothing.
+        """
+        pass
+
+    def get_global_state(self):
+        """
+        Gets the global state which should be a list of dictionaries
+        """
+        timeout = time.time() + 5
+        tx_receipt = None
+        while time.time() < timeout:
+            try:
+                tx_receipt = requests.get(
+                    "http://localhost:{0}/state".format(self.port))
+                tx_receipt.raise_for_status()
+                if tx_receipt:
+                    break
+            except (UnboundLocalError, requests.exceptions.ConnectionError) as e:
+                logging.info("HTTP GET error, got: {0}".format(e))
+                continue
+        logging.info("global state:{}".format(tx_receipt.json()))
+        return tx_receipt.json()
+
+    def get_diffs(self, old_state: list, new_state: dict) -> list:
+        """
+        Iterate through oldState and newState to see any differences
+        Take action based on the differences
+        """
+        tx_diffs = [tx for tx in new_state.get('messages') if (
+                        tx not in old_state)]
+        return tx_diffs
+
+    def get_state_diffs(self, event_filter: object) -> list:
+        """
+        Gets state, then finds diffs, then sets state of blockchain.
+        """
+        new_state = self.get_global_state()
+        if new_state:
+            state_diffs = self.get_diffs(self.state, new_state)
+            filtered_diffs = [tx for tx in state_diffs if event_filter(tx)]
```

Comment:
Nitpick: `filter` functions are faster than list comprehensions, so if `state_diffs` is large enough, I'd recommend using `filter`.

Reply:
wtf is a filter